pycloudlib/lxd/cloud.py

```diff
@@ -448,11 +459,13 @@ def _image_info(self, image_id, is_vm=False):
 
         return image_info
 
-    def image_serial(self, image_id):
+    def image_serial(self, image_id, is_vm=False, **kwargs):
```

Comment:
I'm a little concerned about adding cloud-specific keyword arguments to these APIs. In particular, doing so with `is_vm` means that code that's written against the pure `BaseCloud` interface needs special handling code to deal with LXD VM instances.

Consider this simple launch example (inspired by cloud-init's integration testing):

```py
def do_launch(cloud: BaseCloud, image_id: str) -> BaseInstance:
    image_id = cloud.released_image(image_id)
    instance = cloud.launch(image_id, wait=False)
    instance.wait(raise_on_cloudinit_failure=False)
    return instance
```

This can be used to launch EC2 or LXD containers:

```py
do_launch(pycloudlib.ec2.cloud.EC2("our_tag"))
do_launch(pycloudlib.lxd.cloud.LXD("our_tag"))
```

but there's no way to launch a LXD VM using it. It would need modification to take `kwargs` (because it doesn't make sense to have `is_vm` on a function which handles non-LXD clouds too), and of course its callers would also need special handling to pass in `kwargs={"is_vm": True}` iff they were launching a LXD VM. This would propagate all the way out through APIs which otherwise could just be passing around a `BaseCloud` instance.

It feels to me like we could modify this abstraction to improve matters. We could have `is_vm` be an instance attribute, and so passed at instantiation time; this would mean that instances of `BaseCloud` could still be passed around. However, that still doesn't allow us to pass around `BaseCloud` _subclasses_ and instantiate them at will: you have to carry `(BaseCloudSubclass, {"kwargs": "to instantiate it with"})` around to be able to do that.

So I think that perhaps the better abstraction is to have separate classes for `LXDContainer` and `LXDVirtualMachine`. This would allow them both to be treated as equal, first-class "clouds", and would enable use cases like those described above. These would share a lot of their implementation, of course. This also follows the abstraction that James settled on in the cloud-init integration testing: we have a `LxdContainerCloud` (which wraps `pycloudlib.lxd.cloud.LXD`), not a `LxdCloud`.

What do folks think?

Reply:
I like this proposal. Even though much of the implementation is the same, I think it makes sense to think of them as two separate "clouds". It also allows for further divergence in implementation.

Reply:
As further background, LXD has `instance_type` and `type` separate in its API: https://linuxcontainers.org/lxd/docs/master/rest-api#post-. This indicates, to me, that container vs VM is a more fundamental difference in their data model than "instance type" is in most other "clouds", which I think supports the use of separate classes in our own data model.

Reply:
I think it is a great idea, since you totally right that the `is_vm` modification is breaking the main abstraction of pycloudlib (Which I missed when implementing that). I can update the code without an issue

Reply:
@OddBloke I was originally likening LXD deployment of vms vs containers as sort of like ec2 managing different instance_types. Generic LXD cloud provides an wrapper around the lxc command set which is used by both lxc --vm and container based instance launches.

TLDR: +1 on  LXDContainer/LXDVirtualmachine subclasses.

I spent a good hour trying to justify a common single LXD cloud as it is abstracting the common CLI for vm and container deployments,  but a couple of things swayed me to subclasses (other than your well thought out explanation)
   - vms vs containers requiring special image filtering, search and launch knowledge,
   - vms require specific lxc profile love during setup and deployment
   - method signature bloat adding in LXCCloud is_vm param into multiple LXDCloud methods such as init, __init__, launch and image* methods indicating enough of a smell that this is better separated from a common class.

I also thought (incorrectly) that `--vm` was nearly a simple alias for `--type virtual-machine`. 

But, when looking at `lxc --type` vs `lxc --vm`, as Dan pointed out in MM, one is not --vm is not actually an alias --type virtual-machine. The two params provide [separate fields](https://github.com/lxc/lxd/blob/master/lxc/init.go#L208-L216) to [lxd's API](https://linuxcontainers.org/lxd/docs/master/rest-api#post-) that direct the underlying cloud platform chosen for the instance deployment.

Through @lucasmoura's good work, we are seeing enough discrepancies in these underlying platforms and how to query images as well as which profiles to use that probably warrant the subclasses suggestion.
Ultimately, under the hood below lxc's CLI, they are unique cloud platforms and probably should be treated as such.